src/Storages/S3Queue/S3QueueFilesMetadata.cpp

```diff
@@ -28,326 +42,875 @@ namespace
 }
 
-S3QueueFilesMetadata::S3FilesCollection S3QueueFilesMetadata::getUnorderedProcessedFiles()
+S3QueueFilesMetadata::ProcessingNodeHolder::~ProcessingNodeHolder()
 {
-    auto zookeeper = storage->getZooKeeper();
-    S3QueueProcessedCollection processed_collection(max_set_size, max_set_age_sec);
-    processed_collection.parse(zookeeper->get(zookeeper_processed_path));
-    return processed_collection.getFileNames();
+    if (!removed)
+        remove();
 }
 
-S3QueueFilesMetadata::S3FilesCollection S3QueueFilesMetadata::getProcessedFailedAndProcessingFiles()
+bool S3QueueFilesMetadata::ProcessingNodeHolder::remove(Coordination::Requests * requests, Coordination::Responses * responses)
 {
-    S3FilesCollection processed_and_failed_files = getFailedFiles();
-    switch (mode)
+    if (removed)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Processing node is already removed");
+
+    LOG_TEST(log, "Removing processing node {} ({})", zk_node_path, path);
+
+    try
     {
-        case S3QueueMode::UNORDERED:
+        if (!zk_client->expired())
         {
-            processed_and_failed_files.merge(getUnorderedProcessedFiles());
-            break;
+            /// Is is possible that we created an ephemeral processing node
+            /// but session expired and someone other created an ephemeral processing node.
+            /// To avoid deleting this new node, check processing_id.
+            std::string res;
+            Coordination::Stat stat;
+            if (zk_client->tryGet(zk_node_path, res, &stat))
+            {
+                auto node_metadata = NodeMetadata::fromString(res);
+                if (node_metadata.processing_id == processing_id)
+                {
+                    if (requests)
+                    {
+                        requests->push_back(zkutil::makeRemoveRequest(zk_node_path, stat.version));
+                        auto code = zk_client->tryMulti(*requests, *responses);
+                        removed = code == Coordination::Error::ZOK;
+                    }
+                    else
+                    {
+                        zk_client->remove(zk_node_path);
+                        removed = true;
+                    }
+                    return removed;
+                }
+                else
+                    LOG_WARNING(log, "Cannot remove {} since processing id changed: {} -> {}",
+                                zk_node_path, processing_id, node_metadata.processing_id);
+            }
+            else
+                LOG_DEBUG(log, "Cannot remove {}, node doesn't exist, "
+                          "probably because of session expiration", zk_node_path);
+
+            /// TODO: this actually would mean that we already processed (or partially processed)
+            /// the data but another thread will try processing it again and data can be duplicated.
+            /// This can be solved via persistenly saving last processed offset in the file.
```

Comment:
Saving offsets won't work for formats with metadata like Parquet/ORC/Arrow/Avro, because we cannot just start reading from some offset in file